core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java

```diff
@@ -719,6 +555,257 @@ private void discoverFields(XClass clazz, Set<XClass> examinedClasses, Map<Integ
+            }
+            if (field.getAnnotation(ProtoField.class) != null) {
+               throw new ProtoSchemaBuilderException("The @ProtoUnknownFieldSet and @ProtoField annotations cannot be used together: " + field);
+            }
+            unknownFieldSetField = field;
+         } else {
+            ProtoField annotation = field.getAnnotation(ProtoField.class);
+            if (annotation != null || implicitFields) {
+               validateField(clazz, field);
+               int number = annotation != null ? getNumber(annotation, field) : position;
+               String fieldName = getName(annotation, field);
+               Type protobufType = defaultType(annotation, field.getType());
+               boolean isArray = isArray(field.getType(), protobufType);
+               boolean isRepeated = isRepeated(field.getType(), protobufType);
+               boolean isRequired = annotation != null && annotation.required();
+               if (isRequired && protoSchemaGenerator.syntax() != ProtoSyntax.PROTO2) {
+                  throw new ProtoSchemaBuilderException("Field '" + fieldName + "' of " + clazz.getCanonicalName() + " cannot be marked required when using \"" + protoSchemaGenerator.syntax() + "\" syntax, while processing " + this.protoSchemaGenerator.generator);
+               }
+               boolean isMap = isMap(field.getType());
+               if (isMap && protoSchemaGenerator.syntax() == ProtoSyntax.PROTO2) {
+                  throw new ProtoSchemaBuilderException("Field '" + fieldName + "' of " + clazz.getCanonicalName() + " of type map is not supported when using \"" + protoSchemaGenerator.syntax() + "\" syntax, while processing " + this.protoSchemaGenerator.generator);
+               }
+               if (isRepeated && isRequired) {
+                  throw new ProtoSchemaBuilderException("Repeated field '" + fieldName + "' of " + clazz.getCanonicalName() + " cannot be marked required.");
+               }
+               XClass javaType = getJavaTypeFromAnnotation(annotation);
+               if (javaType == typeFactory.fromClass(void.class)) {
+                  javaType = isRepeated ? field.determineRepeatedElementType() : field.getType();
+               }
+               if (javaType == typeFactory.fromClass(byte[].class) && protobufType == Type.MESSAGE) {
+                  // MESSAGE is the default and stands for 'undefined', we can override it with a better default
+                  protobufType = Type.BYTES;
+               }
+               if (!javaType.isArray() && !javaType.isPrimitive() && javaType.isAbstract() && !javaType.isEnum()) {
+                  throw new ProtoSchemaBuilderException("The type " + javaType.getCanonicalName() + " of field '" + fieldName + "' of " + clazz.getCanonicalName() + " should not be abstract, while processing " + this.protoSchemaGenerator.generator);
```

Comment:
We throw the same exception in two places. We should either split the duplicated code into a method, or at least create the exception via the Log.

Reply:
I've promoted a few logs to `Log`, including this one.